packages/medusa/src/api/routes/admin/inventory-items/transaction/create-inventory-item.ts

```diff
@@ -0,0 +1,174 @@
+import {
+  DistributedTransaction,
+  TransactionHandlerType,
+  TransactionOrchestrator,
+  TransactionPayload,
+  TransactionState,
+  TransactionStepsDefinition,
+} from "../../../../../utils/transaction"
+import { ulid } from "ulid"
+import { EntityManager } from "typeorm"
+import { IInventoryService } from "../../../../../interfaces"
+import {
+  ProductVariantInventoryService,
+  ProductVariantService,
+} from "../../../../../services"
+import { CreateProductVariantInput } from "../../../../../types/product-variant"
+import { InventoryItemDTO } from "../../../../../types/inventory"
+import { ProductVariant } from "../../../../../models"
+import { MedusaError } from "medusa-core-utils"
+
+enum actions {
+  createInventoryItem = "createInventoryItem",
+  attachInventoryItem = "attachInventoryItem",
+}
+
+const flow: TransactionStepsDefinition = {
+  next: {
+    action: actions.createInventoryItem,
+    saveResponse: true,
+    next: {
+      action: actions.attachInventoryItem,
+      noCompensation: true,
+    },
+  },
+}
+
+const createInventoryItemStrategy = new TransactionOrchestrator(
+  "create-variant-with-inventory",
+  flow
+)
+
+type InjectedDependencies = {
+  manager: EntityManager
+  productVariantService: ProductVariantService
+  productVariantInventoryService: ProductVariantInventoryService
+  inventoryService?: IInventoryService
+}
+
+type CreateInventoryItemInput = {
+  sku?: string
+  hs_code?: string
+  weight?: number
+  length?: number
+  height?: number
+  width?: number
+  origin_country?: string
+  mid_code?: string
+  material?: string
+  metadata?: Record<string, unknown>
+}
+
+export const createInventoryItemTransaction = async (
+  dependencies: InjectedDependencies,
+  variantId: string,
+  input: CreateInventoryItemInput
+): Promise<DistributedTransaction> => {
+  const {
+    manager,
+    productVariantService,
+    inventoryService,
+    productVariantInventoryService,
+  } = dependencies
+
+  const inventoryServiceTx = inventoryService?.withTransaction(manager)
```

Comment:
we have decided to do not share DB transaction with modules. So you can remove the `withTransaction` everywhere and let the TO handle the compensation.